Make LXC works

tieong opened the pull request:
Hey there,

First time I do a Nix related PR, so would appreciate pointers to do things better.

This PR allow us to create privileged and unprivileged LXC containers, the only thing we are missing here is apparmor support which should not be too hard to enable.

I dug a little into this and we just need to wrap the lxc package so that it can finds apparmor_parser, also it does hardcode stuff such as /usr/bin/lxc-start so we need to patch that as well, and the newuidmap/newgidmap binaries

```
INFO     idmap_utils - ../src/lxc/idmap_utils.c:lxc_map_ids:165 - newuidmap binary is missing
INFO     idmap_utils - ../src/lxc/idmap_utils.c:lxc_map_ids:171 - newgidmap binary is missing
DEBUG    idmap_utils - ../src/lxc/idmap_utils.c:lxc_map_ids:186 - No newuidmap and newgidmap binary found. Trying to write directly with euid 0
```

Then it just needs more docs and examples on how to set this, the rest of the pain points I had are in the commit messages.

But it's mostly about how we handle lxc includes, the lxc package and run_command sed problems.

tieong committed:
pve-lxc-syscalld: Add package and fix execstart

pve-lxc-syscalld: Call pve-lxc-syscalld package

pve-lxc-syscalld: Add package and fix execstart

tieong committed:
pve-ha-manager: Make pct cli works

tieong committed:
pve-common: Substitute diff and ip

We already propagate diff and ip binaries via propagatedBuildInputs
but this does not seems to works everywhere.

Take the iface_set_master subroutine in pve-common/src/PVE/Network.pm

I thought it would know how to find ip but it doesn't for some
reason. Is it because it is in an eval block and the context is
somewhat different?

tieong committed:
pve-common: Don't hardcode /bin/bash in run_command

tieong committed:
container: Enable container.nix

tieong committed:
pve-manager: Allow LXC creation from the web ui

Without these it cannot untar archives

tieong committed:
pve-container: Remove dsa from ssh key generation

https://forum.proxmox.com/threads/lxc-container-creation-failing-dsa-ssh-key-generation-error.155813/

tieong committed:
proxmox-ve: Add an execstartpre to mkdir /run/pve

tieong committed:
container.nix: Add lxc-monitord service and pve lxc slice

tieong committed:
lxc: Add lxc package

This was taken straight out of the nixpkgs definition, there is
probably a better way than doing this.

This was done because there were hardcoded stuff in the lxc package
which I needed to replace and it was easier to refer to this package
from pve-container.

I only made 2 modifications compared to the nixpkgs all in the
postInstall section:

I add this at the end, because it was complaining about missing coreutils:

    sed -i $out/libexec/lxc/lxc-containers \
      -e "s|touch|${coreutils}/bin/touch|" \
      -e "s|rm|${coreutils}/bin/rm|"

And I removed this line:

     substituteInPlace $out/share/lxc/config/common.conf --replace-fail "$out/share" "/run/current-system/sw/share"

/run/current-system/sw/share was empty, not sure how it's supposed to
be populated.

tieong committed:
pve-container: Added missings deps and fixed harcoded paths

tieong committed:
container.nix: Fixed pve-container services

To use proper types, for example null wasn't getting rendered in the
final service file, needed to add it between quotes.

tieong committed:
container.nix: Add lxc and lxc-net services

Added these services because that's how it's done in a vanilla proxmox
install, but I think pve-container service being presents are enough
IIRC, we could try removing the wants lxc.

Also instead of defining these here, maybe we could leverage the
builtin virtualisation.lxc? I did not do so because it was
quicker/easier with these services present directly.

tieong committed:
pve-container: Fixed lxc includes

LXC config file that gets generated needs to include a lot of stuff.

First is LXCFS so that containers actually know what are their true
uptime for example not the host.

Second seccomp rules

And the rest are common config, note that ideally it also wants distro
specific files, these are not packaged for nix yet.

https://github.com/lxc/lxc-templates/tree/main/config

tieong commented:
Also this might be a good opportunity to decide how we handle container templates download and gpg keys, so that we can use the pveam download/list/available command and download from the gui.

JulienMalka commented:
Ideally, given that this is a big piece of functionality, and given that I do not personally use this functionality, to make sure that this doesn't break with updates without anyone realizing, we need a good coverage of it with NixOS tests.

tieong committed:
pve-lxc-syscalld: Get rid of cranelib

tieong committed:
tests: Add test for containers

tieong committed:
pve-manager: Make pveam update command work

The patch is necessary because gpgv complains if the keyring is in
armor format instead of binary.

Latest version of sequoia do not allow to produce binary output so
this will be needed in the future as well.

tieong committed:
lxc: Use setuid versions of newgidmap/newuidmap, fixed apparmor path

tieong committed:
Merge branch 'main' into lxc

tieong committed:
pve-common: Removed /bin/bash path patch

This should already be replaced via a sed, furthermore substituteAll
has been deprecated and throw an error now.

tieong committed:
pve-manager: fixed patch

tieong committed:
pve-storage: Update version to 9.0.1

JulienMalka commented:
I am currently very busy with a deadline, but will review and help you merge this PR after july 18.

max-zelinski commented:
any updates on getting this merged by chance?

Thrawn7 commented:
Seems like this got pretty far. Any updates on getting LXCs working?
